File: backend/src/services/discord-notifier.js
const defaults = {
  username: 'EnderScope',
  title: 'New Server Found: {ip}:{port}',
  description:
    '{motd}\nVersion: {version}\nPlayers: {players_online}/{players_max}\nWhitelist: {whitelist}\nSource: {source}',
  color: '#8b5cf6',
};

function normalizeColor(value) {
  const raw = String(value || '').trim().replace(/^#/, '');
  if (/^[0-9a-fA-F]{6}$/.test(raw)) {
    return parseInt(raw, 16);
  }
  return parseInt(defaults.color.replace('#', ''), 16);
}

function getTemplateContext(server) {
  const whitelist =
    server.whitelist === false ? 'Open' : server.whitelist ? 'Whitelisted' : 'Unknown';

  return {
    ip: server.ip || 'unknown',
    port: String(server.port || 25565),
    version: server.version || 'unknown',
    motd: server.motd || '-',
    players_online: String(server.players_online || 0),
    players_max: String(server.players_max || 0),
    whitelist,
    source: server.source || 'unknown',
  };
}

function applyTemplate(template, server) {
  const context = getTemplateContext(server);
  return String(template || '').replace(/\{([a-z_]+)\}/gi, (_, token) => context[token] ?? '');
}

export async function notifyDiscord(settings, server) {
  const webhookUrl =
    typeof settings === 'string' ? settings : String(settings?.discord_webhook_url || '');

  if (!webhookUrl.startsWith('https://discord.com/api/webhooks/')) {
    return false;
  }

  const whitelistValue =
    server.whitelist === false ? 'Open' : server.whitelist ? 'Whitelisted' : 'Unknown';

  const payload = {
    username: String(settings?.discord_webhook_username ?? defaults.username).slice(0, 80),
    embeds: [
      {
        title: applyTemplate(settings?.discord_webhook_title ?? defaults.title, server).slice(0, 256),
        description: applyTemplate(
          settings?.discord_webhook_description ?? defaults.description,
          server
        ).slice(0, 4096),
        color: normalizeColor(settings?.discord_webhook_color),
        fields: [
          { name: 'IP:Port', value: `${server.ip}:${server.port}`, inline: true },
          { name: 'Version', value: server.version || 'unknown', inline: true },
          {
            name: 'Players',
            value: `${server.players_online || 0}/${server.players_max || 0}`,
            inline: true,
          },
          { name: 'Whitelist', value: whitelistValue, inline: true },
          { name: 'Source', value: server.source || 'unknown', inline: true },
        ],
      },
    ],
  };

  const response = await fetch(webhookUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });

  return response.ok;
}

File: backend/legacy_python/core/discord_notifier.py
"""Discord webhook notifier."""
from __future__ import annotations

import logging
from typing import Sequence

import requests

from backend.core.models import ServerResult

logger = logging.getLogger(__name__)

_MAX_EMBED_FIELDS = 25


class DiscordNotifier:
    """Send server find notifications via a Discord webhook."""

    def __init__(self, webhook_url: str, timeout: float = 10.0) -> None:
        self._url = webhook_url
        self._timeout = timeout

    def is_configured(self) -> bool:
        return bool(self._url and self._url.startswith("https://discord.com/api/webhooks/"))

    def notify_single(self, server: ServerResult) -> bool:
        """Send a single server result embed."""
        if not self.is_configured():
            logger.warning("Discord webhook not configured — skipping notification.")
            return False

        whitelist_str = (
            "\u2705 Open" if server.whitelist is False
            else ("\u274c Whitelisted" if server.whitelist else "\u2753 Unknown")
        )

        embed = {
            "title": f"\U0001f30d New Server Found: {server.address}",
            "color": 0x00FF88 if not server.whitelist else 0xFF4444,
            "fields": [
                {"name": "IP:Port", "value": server.address, "inline": True},
                {"name": "Version", "value": server.version or "unknown", "inline": True},
                {"name": "MOTD", "value": server.motd[:256] or "—", "inline": False},
                {"name": "Players", "value": f"{server.players_online}/{server.players_max}", "inline": True},
                {"name": "Whitelist", "value": whitelist_str, "inline": True},
                {"name": "Source", "value": server.source, "inline": True},
            ],
        }
        return self._post({"embeds": [embed]})

    def notify_batch(self, servers: Sequence[ServerResult]) -> bool:
        """Send a summary embed for multiple servers."""
        if not self.is_configured():
            return False

        fields = [
            {
                "name": s.address,
                "value": "`{}` \u2014 {}".format(s.version, "\u2705" if not s.whitelist else "\u274c"),
                "inline": True,
            }
            for s in servers[:_MAX_EMBED_FIELDS]
        ]
        embed = {
            "title": f"\U0001f4e1 ServerBuster found {len(servers)} server(s)",
            "color": 0x5865F2,
            "fields": fields,
        }
        return self._post({"embeds": [embed]})

    def _post(self, payload: dict) -> bool:
        try:
            response = requests.post(
                self._url, json=payload, timeout=self._timeout
            )
            response.raise_for_status()
            return True
        except requests.RequestException as exc:
            logger.error("Discord notification failed: %s", exc)
            return False
